rfd/0155-scoped-webauthn-credentials.md

```diff
@@ -0,0 +1,197 @@
+---
+authors: Brian Joerger (bjoerger@goteleport.com)
+state: draft
+---
+
+# RFD 155 - Scoped Webauthn Credentials
+
+## Required Approvers
+
+- Engineering: @rosstimothy && @codingllama
+- Product: @klizhentas || @xinding33
+- Security: @jentfoo
+
+## What
+
+Add and enforce scope for webauthn credentials.
+
+## Why
+
+Today, when a client performs a webauthn ceremony, the resulting webauthn
+credentials can be used to perform actions other than what the client initially
+intended.
+
+For example, webauthn credentials intended for per-session MFA could be used
+instead for login. This means that if an attacker gets a hold of a user's
+webauthn credentials before the client consumes them, an attacker with the
+user's password could login as the user. Adding scope to the webauthn
+credentials would prevent the attacker from using the stolen webauthn
+credentials freely.
+
+Adding scope also enables us to allow webauthn credentials to be reused within
+specific scopes. For example, for admin actions it may make sense to allow the
+client to reuse a single webauthn credential for a string of requests that are
+functionally a single action, such as creating a new user and their corresponding
+invite token.
+
+## Details
+
+### Security
+
+Enforcing scope for webauthn credentials strictly improves security, as it is a
+purely restrictive change that prevents certain attack vectors, as described
+in [Why](#why).
+
+On the other hand, allowing clients to reuse webauthn credentials could have
+negative security implications if not handled carefully. Specifically, the
+client and server must clearly define:
+
+1. The webauthn credential's scope - client provided
+2. The number of reuses permitted - client provided
+3. The expiration of the credentials - server enforced (5 minutes)
+
+Additionally, reusing credentials will be limited to specific scopes where it
```

Comment:
Should we mention how challenges are supplied during reuse?

Should we mention in the design that reuse and scopes are only explicitly allowed for certain RPCs, as decided by the server impl, or is that a given?

Reply:
> Should we mention how challenges are supplied during reuse?

I think this is sort of out of scope. The credentials can be supplied in different ways depending on the transport. Admin actions will resupply the same MFA response through the gRPC call opts.

> Should we mention in the design that reuse and scopes are only explicitly allowed for certain RPCs, as decided by the server impl, or is that a given?

Do you mean that we should limit reuse to specific admin actions? I was planning to keep this out of scope, leaving it up to the client to determine if reuse is acceptable. I think we can forbid reuse for a few specific endpoints.

Reply:
Re challenge transport: I think a nod to how it works is good to have for folks reading this in the future. As of now it's only implied and you'd need to know about other RFDs/features.

Re reuse to specific actions: I think this depends on how you want to implement this, but I was thinking about making reuse explicit for the RPCs that need them, as in they need some sort of code to "fetch" verify a reusable challenge with a certain scope. This makes it harder for unintended reuse to happen, at the cost of maintenance/more code.